File: CS/TreeViewReport/TreeReport.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using DevExpress.XtraReports.UI;
using System.Data;
using System.Linq;
using System.Collections.Generic;

namespace TreeViewReport {
    public partial class TreeReport : DevExpress.XtraReports.UI.XtraReport {
        List<ParentNode> nodesList;
        DataHelper helper;
        Form1 parentForm;
        public DataHelper Helper { 
            get {
                return this.helper;
            }
        }
        public TreeReport() {
            InitializeComponent();
            this.ParametersRequestSubmit += TreeReport_ParametersRequestSubmit;
        }

        public TreeReport(Form1 parentForm) {
            InitializeComponent();
            this.parentForm = parentForm;
            this.ParametersRequestSubmit += TreeReport_ParametersRequestSubmit;
        }
        

        void TreeReport_ParametersRequestSubmit(object sender, DevExpress.XtraReports.Parameters.ParametersRequestEventArgs e) {
            int level = Convert.ToInt32(this.Parameters["Level"].Value);
            helper = new DataHelper(level);
            nodesList  = helper.NodesList;
            this.DataSource = nodesList;
            this.CreateDocument();
        }        

        private void xrSubreport1_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            XRSubreport subReport = sender as XRSubreport;                                    
            ChildReport childReport = new ChildReport();
            childReport.Parameters["Level"].Value = this.Parameters["Level"].Value;
            childReport.DataSource = nodesList;
            subReport.ReportSource = childReport;
            parentForm.treeList1.DataSource = helper.ReportDataSource;
            parentForm.treeList1.ExpandAll();
        }
    }
}

File: VB/TreeViewReport/TreeReport.vb
Imports System
Imports System.Drawing
Imports System.ComponentModel
Imports DevExpress.XtraReports.UI
Imports System.Collections.Generic

Namespace TreeViewReport

    Public Partial Class TreeReport
        Inherits XtraReport

        Private nodesList As List(Of ParentNode)

        Private helperField As DataHelper

        Private parentForm As Form1

        Public ReadOnly Property Helper As DataHelper
            Get
                Return helperField
            End Get
        End Property

        Public Sub New()
            InitializeComponent()
            AddHandler ParametersRequestSubmit, AddressOf TreeReport_ParametersRequestSubmit
        End Sub

        Public Sub New(ByVal parentForm As Form1)
            InitializeComponent()
            Me.parentForm = parentForm
            AddHandler ParametersRequestSubmit, AddressOf TreeReport_ParametersRequestSubmit
        End Sub

        Private Sub TreeReport_ParametersRequestSubmit(ByVal sender As Object, ByVal e As DevExpress.XtraReports.Parameters.ParametersRequestEventArgs)
            Dim level As Integer = Convert.ToInt32(Parameters("Level").Value)
            helperField = New DataHelper(level)
            nodesList = helperField.NodesList
            DataSource = nodesList
            CreateDocument()
        End Sub

        Private Sub xrSubreport1_BeforePrint(ByVal sender As Object, ByVal e As CancelEventArgs)
            Dim subReport As XRSubreport = TryCast(sender, XRSubreport)
            Dim childReport As ChildReport = New ChildReport()
            childReport.Parameters("Level").Value = Parameters("Level").Value
            childReport.DataSource = nodesList
            subReport.ReportSource = childReport
            parentForm.treeList1.DataSource = helperField.ReportDataSource
            parentForm.treeList1.ExpandAll()
        End Sub
    End Class
End Namespace
